lib/apiservers/engine/backends/container_proxy.go

```diff
@@ -317,6 +324,104 @@ func (c *ContainerProxy) CreateExecTask(handle string, config *types.ExecConfig)
+	// to be more accurate.
+	params := tasks.NewInspectParamsWithContext(ctx).WithConfig(config)
+	resp, err := c.client.Tasks.Inspect(params)
+	if err != nil {
+		switch err := err.(type) {
+		case *tasks.InspectNotFound:
+			// These error types may need to be expanded. NotFoundError does not fit here.
+			op.Errorf("received a TaskNotFound error during task inspect: %s", err.Payload.Message)
+			return nil, ConflictError(fmt.Sprintf("container (%s) has been poweredoff", cid))
+		case *tasks.InspectInternalServerError:
+			op.Errorf("received an internal server error during task inspect: %s", err.Payload.Message)
+			return nil, InternalServerError(err.Payload.Message)
+		default:
+			// right now Task inspection in the portlayer does not return a conflict error
+			return nil, InternalServerError(err.Error())
+		}
+	}
+	return resp.Payload, nil
+}
+
+func (c *ContainerProxy) BindTask(op trace.Operation, handle string, eid string) (string, error) {
+	defer trace.End(trace.Begin(fmt.Sprintf("handle(%s), eid(%s)", handle, eid)))
+
+	bindconfig := &models.TaskBindConfig{
+		Handle: handle,
+		ID:     eid,
+	}
+	bindparams := tasks.NewBindParamsWithContext(ctx).WithConfig(bindconfig)
+
+	// call Bind with bindparams
+	resp, err := c.client.Tasks.Bind(bindparams)
+	if err != nil {
+		switch err := err.(type) {
+		case *tasks.BindNotFound:
+			op.Errorf("received TaskNotFound error during task bind: %s", err.Payload.Message)
+			return "", NotFoundError("container (%s) has been poweredoff")
+		case *tasks.BindInternalServerError:
+
+			op.Errorf("received unexpected error attempting to bind task(%s) for handle(%s): %s", eid, handle, err.Payload.Message)
+			return "", InternalServerError(err.Payload.Message)
+		default:
+			op.Errorf("received unexpected error attempting to bind task(%s) for handle(%s): %s", eid, handle, err.Error())
+			return "", InternalServerError(err.Error())
+		}
+
+	}
+
+	respHandle, ok := resp.Payload.Handle.(string)
+	if !ok {
+		op.Errorf("Unable to marshal string object from BindTask response for handle(%s) on eid(%s)", handle, eid)
+		// TODO: perhaps a better error message here?
+		return "", InternalServerError("An unknown error occurred during the handling of this request")
+	}
+
+	return respHandle, nil
+}
+
+func (c *ContainerProxy) WaitTask(op trace.Operation, cid string, cname string, eid string) error {
+	handle, err := c.Handle(cid, cname)
+	if err != nil {
+		return err
+	}
+
+	// wait the Task to start
+	config := &models.TaskWaitConfig{
+		Handle: handle,
+		ID:     eid,
+	}
+
+	params := tasks.NewWaitParamsWithContext(ctx).WithConfig(config)
+	_, err = c.client.Tasks.Wait(params)
+	if err != nil {
+		switch err := err.(type) {
+		case *tasks.WaitNotFound:
+			return InternalServerError("the Container(%s) has been shutdown during execution ofthe exec operation")
+		case *tasks.WaitPreconditionRequired:
+			return InternalServerError("container(%s) must be powered on in order perform the desired exec operation")
```

Comment:
`container must be running to perform an exec` - short is sweet, we've got way too many words adding no value in a lot of these messages.